File: src/cookieSecurity.ts
/**
 * cookieSecurity.ts — Set-Cookie attribute audit. T76.
 *
 * Reads the response's `Set-Cookie` header(s) from the captured
 * top-level navigation response and surfaces cookies that lack
 * the modern security attributes:
 *
 *   * `Secure` — cookie is only sent over https. Required on
 *     https pages; absence allows the cookie to leak over an
 *     http downgrade (MITM-able even briefly).
 *   * `SameSite` — `Strict` blocks the cookie from cross-site
 *     fetches entirely; `Lax` allows top-level GETs but blocks
 *     POSTs and sub-resource fetches; `None` (with `Secure`)
 *     opts back in to all cross-site sends. Missing → modern
 *     browsers default `Lax` but old clients default to no
 *     restriction, allowing CSRF.
 *   * `HttpOnly` — JS can't read the cookie via `document.cookie`.
 *     Critical for session cookies — XSS can't steal them
 *     without it.
 *
 * Findings:
 *
 *   - cookie.no-secure                strict (https only)
 *     Cookie set without `Secure` on an https page.
 *
 *   - cookie.no-samesite              warn
 *     Cookie set without `SameSite` attribute. Modern browsers
 *     default `Lax` but old clients leave it unrestricted.
 *
 *   - cookie.session-no-httponly      warn
 *     Cookie name looks session-like (matches /sess|sid|auth|
 *     token|jwt/i) AND lacks `HttpOnly`. XSS attacker can read
 *     and exfiltrate.
 *
 *   - cookie.samesite-none-no-secure  strict
 *     `SameSite=None` set without `Secure`. Browsers REJECT
 *     this combination — the cookie isn't stored at all.
 *
 * Out of scope:
 *   * http pages — Secure attribute can't apply, no-samesite
 *     warn still fires but no-secure does not.
 *   * Localhost — same exemption family as hsts/xFrameOptions.
 *
 * Reads from the same `topLevelResponseHeaders` Map as
 * hsts + xFrameOptions + referrerPolicy. This is the FOURTH
 * response-header detector — per docs/DETECTORS.md note from
 * cycle 17, this is the right point to extract a generic
 * helper. Deliberately NOT extracted here yet — see the
 * "TODO" in DOGFOOD_RUNS.md cycle 19 — the per-cookie shape
 * is genuinely different from the per-header shape (one
 * response can have MULTIPLE Set-Cookie headers, each with
 * its own attributes), so a generic helper would be
 * shoe-horning. Pattern-extraction deferred to a sibling
 * cycle.
 */

export interface CookieSecurityFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface CapturedCookie {
  name: string;
  /** Empty string if not set. */
  sameSite: string;
  hasSecure: boolean;
  hasHttpOnly: boolean;
  /** Raw Set-Cookie line, truncated. */
  raw: string;
}

export interface CookieSecuritySnapshot {
  pageUrl: string;
  pageIsHttps: boolean;
  pageIsLocalhost: boolean;
  cookies: CapturedCookie[];
}

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

/**
 * Parse one Set-Cookie line. Returns null on malformed (no `=`
 * before the first attribute separator). The raw line is
 * preserved on the result for the audit trail.
 *
 * Set-Cookie syntax: `name=value; Attr1; Attr2=val; ...`
 *
 * Browsers tolerate a lot of weirdness here (whitespace, casing,
 * mixed quoting). We normalise attribute names to lowercase.
 */
function parseSetCookie(raw: string): CapturedCookie | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  const parts = trimmed.split(';');
  const head = parts[0].trim();
  const eq = head.indexOf('=');
  if (eq <= 0) return null;
  const name = head.slice(0, eq).trim();
  if (!name) return null;
  let sameSite = '';
  let hasSecure = false;
  let hasHttpOnly = false;
  for (let i = 1; i < parts.length; i++) {
    const a = parts[i].trim();
    if (!a) continue;
    const lower = a.toLowerCase();
    if (lower === 'secure') {
      hasSecure = true;
    } else if (lower === 'httponly') {
      hasHttpOnly = true;
    } else if (lower.startsWith('samesite=')) {
      sameSite = a.slice('samesite='.length).trim();
    } else if (lower === 'samesite') {
      // Bare `SameSite` (no value) — browsers treat as no-attribute,
      // record empty.
      sameSite = '';
    }
  }
  return {
    name,
    sameSite,
    hasSecure,
    hasHttpOnly,
    raw: trimmed.slice(0, 200),
  };
}

/**
 * Extract every Set-Cookie value from a headers map. HTTP
 * historically allows multiple Set-Cookie headers per response;
 * different servers (and Playwright's exposure) collapse them
 * into one comma-separated value OR keep them separate. We
 * accept both — splitting on `\n` first (Playwright's
 * representation when multiple are present), then on commas
 * IF the first header value contains multiple cookies (rare).
 *
 * The comma-split is risky because cookie expires-attr values
 * may contain commas. Solution: only comma-split when we see a
 * known-cookie-attribute keyword on the right side, leaving
 * Expires=... commas alone.
 */
function extractSetCookieLines(
  headers: Record<string, string> | undefined,
): string[] {
  if (!headers) return [];
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === 'set-cookie') {
      // Playwright's `headers()` joins multiple Set-Cookie with
      // newlines. Split on \n first.
      return v.split('\n').map((s) => s.trim()).filter(Boolean);
    }
  }
  return [];
}

export function buildCookieSecuritySnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): CookieSecuritySnapshot {
  const pageIsHttps = pageUrl.startsWith('https://');
  const pageIsLocalhost = isLocalhost(pageUrl);
  const lines = extractSetCookieLines(headers);
  const cookies: CapturedCookie[] = [];
  for (const line of lines) {
    const c = parseSetCookie(line);
    if (c) cookies.push(c);
  }
  return { pageUrl, pageIsHttps, pageIsLocalhost, cookies };
}

/**
 * Heuristic for "looks like a session cookie". Names that
 * commonly carry session tokens. False positives are fine
 * (warn-only); false negatives mean we miss real defects, so
 * be liberal.
 */
function looksLikeSessionCookie(name: string): boolean {
  return /sess|sid|auth|token|jwt|bearer/i.test(name);
}

export function detectCookieSecurityIssues(
  snap: CookieSecuritySnapshot,
): CookieSecurityFinding[] {
  if (snap.pageIsLocalhost) return [];
  if (snap.cookies.length === 0) return [];

  const noSecure: CapturedCookie[] = [];
  const noSameSite: CapturedCookie[] = [];
  const sessionNoHttpOnly: CapturedCookie[] = [];
  const sameSiteNoneNoSecure: CapturedCookie[] = [];

  for (const c of snap.cookies) {
    // SameSite=None requires Secure; browsers reject otherwise.
    if (c.sameSite.toLowerCase() === 'none' && !c.hasSecure) {
      sameSiteNoneNoSecure.push(c);
    }
    if (snap.pageIsHttps && !c.hasSecure) {
      noSecure.push(c);
    }
    if (!c.sameSite) {
      noSameSite.push(c);
    }
    if (looksLikeSessionCookie(c.name) && !c.hasHttpOnly) {
      sessionNoHttpOnly.push(c);
    }
  }

  const out: CookieSecurityFinding[] = [];
  const renderEx = (c: CapturedCookie) => `'${c.name}' raw='${c.raw}'`;

  if (noSecure.length > 0) {
    const examples = noSecure.slice(0, 5).map(renderEx);
    out.push({
      severity: 'strict',
      kind: 'cookie.no-secure',
      detail: `${noSecure.length} cookie(s) set without 'Secure' attribute on this https page. The cookie can be sent over http if the user is briefly downgraded (MITM, network rewrite, mixed-content). Add '; Secure' to every Set-Cookie. Examples: ${examples.join('; ')}`,
      evidence: { count: noSecure.length, examples },
    });
  }
  if (sameSiteNoneNoSecure.length > 0) {
    const examples = sameSiteNoneNoSecure.slice(0, 5).map(renderEx);
    out.push({
      severity: 'strict',
      kind: 'cookie.samesite-none-no-secure',
      detail: `${sameSiteNoneNoSecure.length} cookie(s) set with 'SameSite=None' but no 'Secure' attribute. Browsers REJECT this combination — the cookie isn't stored at all. Either add Secure (and only set on https) or change to 'SameSite=Lax'. Examples: ${examples.join('; ')}`,
      evidence: { count: sameSiteNoneNoSecure.length, examples },
    });
  }
  if (noSameSite.length > 0) {
    const examples = noSameSite.slice(0, 5).map(renderEx);
    out.push({
      severity: 'warn',
      kind: 'cookie.no-samesite',
      detail: `${noSameSite.length} cookie(s) set without 'SameSite' attribute. Modern browsers default 'Lax' (safe) but older clients leave it unrestricted, exposing the user to CSRF. Set 'SameSite=Strict' (or 'Lax' if cross-site GETs are needed). Examples: ${examples.join('; ')}`,
      evidence: { count: noSameSite.length, examples },
    });
  }
  if (sessionNoHttpOnly.length > 0) {
    const examples = sessionNoHttpOnly.slice(0, 5).map(renderEx);
    out.push({
      severity: 'warn',
      kind: 'cookie.session-no-httponly',
      detail: `${sessionNoHttpOnly.length} session-looking cookie(s) (name matches sess/sid/auth/token/jwt/bearer) lack 'HttpOnly'. JS — including injected XSS — can read these via document.cookie and exfiltrate. Add '; HttpOnly'. Examples: ${examples.join('; ')}`,
      evidence: { count: sessionNoHttpOnly.length, examples },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/cookie_security.rs
//! `cookie_security` — Set-Cookie attribute audit.
//!
//! Mirror of `src/cookieSecurity.ts`. Findings:
//!
//!   * `cookie.no-secure`               strict (https only)
//!   * `cookie.no-samesite`             warn
//!   * `cookie.session-no-httponly`     warn (heuristic on name)
//!   * `cookie.samesite-none-no-secure` strict (browsers REJECT)
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};

/// One parsed Set-Cookie line.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CapturedCookie {
    /// Cookie name (left side of the first `=`).
    pub name: String,
    /// `SameSite` attribute value. Empty if absent or bare.
    pub same_site: String,
    /// Whether the `Secure` attribute was set.
    pub has_secure: bool,
    /// Whether the `HttpOnly` attribute was set.
    pub has_http_only: bool,
    /// Raw Set-Cookie line, truncated to 200 chars.
    pub raw: String,
}

/// Captured page state the detector consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CookieSecuritySnapshot {
    /// Page URL.
    pub page_url: String,
    /// True iff the page was loaded over https.
    pub page_is_https: bool,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Parsed cookies from the top-level navigation response.
    pub cookies: Vec<CapturedCookie>,
}

/// Parse one Set-Cookie line. Returns `None` on malformed input
/// (no `=` before the first attribute separator, or empty name).
///
/// Browsers tolerate weird whitespace / casing; we normalise
/// attribute names to lowercase.
fn parse_set_cookie(raw: &str) -> Option<CapturedCookie> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    let mut parts = trimmed.split(';');
    let head = parts.next().unwrap_or("").trim();
    let eq = head.find('=')?;
    if eq == 0 {
        return None;
    }
    let name = head[..eq].trim().to_owned();
    if name.is_empty() {
        return None;
    }
    let mut same_site = String::new();
    let mut has_secure = false;
    let mut has_http_only = false;
    for raw_attr in parts {
        let a = raw_attr.trim();
        if a.is_empty() {
            continue;
        }
        let lower = a.to_ascii_lowercase();
        if lower == "secure" {
            has_secure = true;
        } else if lower == "httponly" {
            has_http_only = true;
        } else if let Some(rest) = lower.strip_prefix("samesite=") {
            // Preserve original case for evidence display.
            let original_prefix_len = "samesite=".len();
            same_site = a[original_prefix_len..].trim().to_owned();
            // If empty after prefix strip (e.g. "SameSite="), keep empty.
            let _ = rest;
        } else if lower == "samesite" {
            // Bare `SameSite` with no value — same as absent.
            same_site.clear();
        }
    }
    let truncated_len = trimmed.len().min(200);
    let raw_truncated: String = trimmed.chars().take(truncated_len).collect();
    Some(CapturedCookie {
        name,
        same_site,
        has_secure,
        has_http_only,
        raw: raw_truncated,
    })
}

/// Extract every Set-Cookie value from a headers map. The
/// upstream wire form joins multiple cookies on `\n`; we split
/// there to recover the per-cookie list.
fn extract_set_cookie_lines<'a, I, K, V>(headers: I) -> Vec<String>
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    for (k, v) in headers {
        if k.as_ref().eq_ignore_ascii_case("set-cookie") {
            return v
                .as_ref()
                .split('\n')
                .map(|s| s.trim().to_owned())
                .filter(|s| !s.is_empty())
                .collect();
        }
    }
    Vec::new()
}

/// Build a snapshot from a captured response-headers map.
pub fn build_cookie_security_snapshot(
    page_url: &str,
    headers: impl IntoIterator<Item = (impl AsRef<str>, impl AsRef<str>)>,
) -> CookieSecuritySnapshot {
    let page_is_https = page_url.starts_with("https://");
    let page_is_localhost = is_localhost(page_url);
    let lines = extract_set_cookie_lines(headers);
    let cookies: Vec<CapturedCookie> = lines
        .into_iter()
        .filter_map(|l| parse_set_cookie(&l))
        .collect();
    CookieSecuritySnapshot {
        page_url: page_url.to_owned(),
        page_is_https,
        page_is_localhost,
        cookies,
    }
}

/// Heuristic: cookie name matches a session-like pattern.
/// False positives are fine (warn-only); false negatives mean
/// we miss real defects, so be liberal.
fn looks_like_session_cookie(name: &str) -> bool {
    let lower = name.to_ascii_lowercase();
    ["sess", "sid", "auth", "token", "jwt", "bearer"]
        .iter()
        .any(|needle| lower.contains(needle))
}

/// Render the first 5 examples as a `; `-joined string.
fn render_examples(cookies: &[&CapturedCookie]) -> String {
    cookies
        .iter()
        .take(5)
        .map(|c| format!("'{}' raw='{}'", c.name, c.raw))
        .collect::<Vec<_>>()
        .join("; ")
}

/// Pure detector: snapshot → findings. No I/O.
pub fn detect_cookie_security_issues(snap: &CookieSecuritySnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }
    if snap.cookies.is_empty() {
        return Vec::new();
    }

    let mut no_secure: Vec<&CapturedCookie> = Vec::new();
    let mut no_same_site: Vec<&CapturedCookie> = Vec::new();
    let mut session_no_http_only: Vec<&CapturedCookie> = Vec::new();
    let mut same_site_none_no_secure: Vec<&CapturedCookie> = Vec::new();

    for c in &snap.cookies {
        if c.same_site.eq_ignore_ascii_case("none") && !c.has_secure {
            same_site_none_no_secure.push(c);
        }
        if snap.page_is_https && !c.has_secure {
            no_secure.push(c);
        }
        if c.same_site.is_empty() {
            no_same_site.push(c);
        }
        if looks_like_session_cookie(&c.name) && !c.has_http_only {
            session_no_http_only.push(c);
        }
    }

    let mut out = Vec::new();

    if !no_secure.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Strict,
            kind: "cookie.no-secure".into(),
            detail: format!(
                "{} cookie(s) set without 'Secure' attribute on this https page. The cookie can be sent over http if the user is briefly downgraded (MITM, network rewrite, mixed-content). Add '; Secure' to every Set-Cookie. Examples: {}",
                no_secure.len(),
                render_examples(&no_secure)
            ),
        });
    }
    if !same_site_none_no_secure.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Strict,
            kind: "cookie.samesite-none-no-secure".into(),
            detail: format!(
                "{} cookie(s) set with 'SameSite=None' but no 'Secure' attribute. Browsers REJECT this combination — the cookie isn't stored at all. Either add Secure (and only set on https) or change to 'SameSite=Lax'. Examples: {}",
                same_site_none_no_secure.len(),
                render_examples(&same_site_none_no_secure)
            ),
        });
    }
    if !no_same_site.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cookie.no-samesite".into(),
            detail: format!(
                "{} cookie(s) set without 'SameSite' attribute. Modern browsers default 'Lax' (safe) but older clients leave it unrestricted, exposing the user to CSRF. Set 'SameSite=Strict' (or 'Lax' if cross-site GETs are needed). Examples: {}",
                no_same_site.len(),
                render_examples(&no_same_site)
            ),
        });
    }
    if !session_no_http_only.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cookie.session-no-httponly".into(),
            detail: format!(
                "{} session-looking cookie(s) (name matches sess/sid/auth/token/jwt/bearer) lack 'HttpOnly'. JS — including injected XSS — can read these via document.cookie and exfiltrate. Add '; HttpOnly'. Examples: {}",
                session_no_http_only.len(),
                render_examples(&session_no_http_only)
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(headers: &[(&str, &str)], url: &str) -> CookieSecuritySnapshot {
        build_cookie_security_snapshot(url, headers.iter().map(|(k, v)| (*k, *v)))
    }

    #[test]
    fn localhost_skipped() {
        let s = snap(&[("Set-Cookie", "sid=abc")], "http://localhost:8000/");
        assert!(detect_cookie_security_issues(&s).is_empty());
    }

    #[test]
    fn no_cookies_clean() {
        let s = snap(&[], "https://example.com/");
        assert!(detect_cookie_security_issues(&s).is_empty());
    }

    #[test]
    fn https_no_secure_is_strict() {
        let s = snap(
            &[("Set-Cookie", "user=joe; SameSite=Lax")],
            "https://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(f
            .iter()
            .any(|x| x.kind == "cookie.no-secure" && x.severity == AxisSeverity::Strict));
    }

    #[test]
    fn http_no_secure_doesnt_fire() {
        // http page → Secure can't apply → no-secure finding suppressed
        let s = snap(
            &[("Set-Cookie", "user=joe; SameSite=Lax")],
            "http://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(!f.iter().any(|x| x.kind == "cookie.no-secure"));
    }

    #[test]
    fn no_samesite_warns() {
        let s = snap(
            &[("Set-Cookie", "user=joe; Secure")],
            "https://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(f.iter().any(|x| x.kind == "cookie.no-samesite"));
    }

    #[test]
    fn samesite_none_without_secure_is_strict() {
        let s = snap(
            &[("Set-Cookie", "tracker=1; SameSite=None")],
            "https://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(f
            .iter()
            .any(|x| x.kind == "cookie.samesite-none-no-secure"
                && x.severity == AxisSeverity::Strict));
    }

    #[test]
    fn session_cookie_without_httponly_warns() {
        let s = snap(
            &[("Set-Cookie", "sid=abc; Secure; SameSite=Lax")],
            "https://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(f.iter().any(|x| x.kind == "cookie.session-no-httponly"));
    }

    #[test]
    fn session_heuristic_matches_jwt_and_token() {
        for name in ["jwt", "auth_token", "MY_SID", "bearer-X"] {
            assert!(looks_like_session_cookie(name), "{name} should match");
        }
    }

    #[test]
    fn non_session_with_no_httponly_doesnt_warn_session() {
        let s = snap(
            &[("Set-Cookie", "prefs=dark; Secure; SameSite=Lax")],
            "https://example.com/",
        );
        let f = detect_cookie_security_issues(&s);
        assert!(!f.iter().any(|x| x.kind == "cookie.session-no-httponly"));
    }

    #[test]
    fn multiple_cookies_join_via_newline_separator() {
        let raw = "a=1; Secure; SameSite=Lax\nsid=xyz; Secure; SameSite=Lax";
        let s = snap(&[("Set-Cookie", raw)], "https://example.com/");
        assert_eq!(s.cookies.len(), 2);
        // sid lacks HttpOnly → session warn
        let f = detect_cookie_security_issues(&s);
        assert!(f.iter().any(|x| x.kind == "cookie.session-no-httponly"));
    }

    #[test]
    fn malformed_set_cookie_is_dropped() {
        // No `=` in head → parser returns None → not in snapshot.
        let s = snap(&[("Set-Cookie", "garbage")], "https://example.com/");
        assert!(s.cookies.is_empty());
    }

    #[test]
    fn case_insensitive_set_cookie_header_lookup() {
        let s = snap(
            &[("SET-COOKIE", "a=1; Secure; SameSite=Lax")],
            "https://example.com/",
        );
        assert_eq!(s.cookies.len(), 1);
    }
}
